app/grandchallenge/algorithms/forms.py

```diff
@@ -1377,3 +1393,60 @@ def clean(self):
         )
 
         return cleaned_data
+
+
+class AlgorithmInterfaceGetOrCreateForm(AlgorithmInterfaceBaseForm):
+    set_as_default = BooleanField(required=False)
+
+    class Meta(AlgorithmInterfaceBaseForm.Meta):
+        fields = (
+            *AlgorithmInterfaceBaseForm.Meta.fields,
+            "set_as_default",
+        )
+
+    def __init__(self, *args, algorithm, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._algorithm = algorithm
+
+        if not self._algorithm.default_interface:
+            self.fields["set_as_default"].initial = True
+
+    def clean(self):
+        cleaned_data = super().clean()
+
+        if (
+            cleaned_data["existing_io"]
+            and AlgorithmAlgorithmInterface.objects.filter(
+                algorithm=self._algorithm,
+                interface=cleaned_data["existing_io"],
+            ).exists()
+        ):
+            raise ValidationError(
+                "Your algorithm already has an interface with these inputs "
+                "and outputs. If you would like to update the 'is_default' "
+                "property of the interface, you can do so by updating the "
+                "existing interface on the interface list."
+            )
+
+        return cleaned_data
+
+    def save(self):
+        io = (
+            self.cleaned_data["existing_io"]
+            if self.cleaned_data["existing_io"]
+            else super().save()
```

Comment:
Is passing around the `existing_io` object necessary? I thought that with the new `create` method the call to `get_existing_interface_for_inputs_and_outputs` would be unnecessary?

Reply:
That is true, but we're doing the check for the existing interface in the base form (that is used in the admin and in this view). Calling `create` in the base form is not what we want because in the admin, we want to throw an error if there is an existing interface rather than create a new one. So then `existing_io` is already there and it would be weird not to use it, right? I could also get rid of the base form and have two seperate, independent forms for the admin and the user-facing view, but there is also the filter on the CI queryset and the check for duplicate interfaces in the base form, so it made sense to have a common base form. What do you think?

Reply:
> in the admin, we want to throw an error if there is an existing interface rather than create a new one. 

Why do we want to do that? I would have thought that the same logic would apply everywhere, that the existing interface is returned if it exists. Do we even need separate admin functionality for this?

I think using the `create` method on the queryset is more aligned with Django, therefore less surprising to future developers. Right now on main we have two forms and a mixin to do common work, but I am unsure why we need different behaviour.

Reply:
I thought it would make sense to raise an error in the admin to make the staff user aware of the fact that there already is an interface with the chosen inputs and outputs, but you're right that we can also just return the existing interface. 

Would you agree that the check for an existing interface on the algorithm that I added to the `clean` function of `AlgorithmInterfaceGetOrCreateForm` is useful? We can also just not create another entry if that is the case, but I thought it would be useful to alert the user that the create form is not meant for updating the "default" interface.

Reply:
> Would you agree that the check for an existing interface on the algorithm that I added to the clean function of AlgorithmInterfaceGetOrCreateForm is useful?

I think it would be more user friendly to handle it here rather than sending them elsewhere.

Reply:
Ok. Then we will end up with two ways of updating the `is_default` property of an algorithm interface (CreateView and  UpdateView), which felt strange to me, but I don't have a strong opinion on this to be honest and am also fine with handling it here.

Reply:
I think that's okay - because these objects are immutable and globally unique the create view and update view are going to be the same thing, right?

Reply:
I have not thought much about the update view. All it should allow a user to do is update the `is_default` property on the through table entry. So I was envisioning a separate form and view for that. Possibly using htmx so that it can be done from the list view, or else a button on a detail view (like the activate algorithm image view).

Reply:
I'm working on using the custom `create()` method rather than the usual `save()` on the `ModelForm`. I think the only way to do that is to override the form's `save` method since we don't want the usual saving to happen: 

```python
 def save(self, *args, **kwargs):
        return AlgorithmInterface.objects.create(
            inputs=self.cleaned_data["inputs"], outputs=self.cleaned_data["outputs"]
        )
```

That works fine for the form in the user-facing view, but it does not work in the admin. Django complains that it needs a `save_m2m` method (which is usually attached to the form in the `save` method when `commit=False`). I don't quite understand why it is throwing this error but it looks like the admin view handles things a little different and breaks when you override the save method. Is there a better way to achieve this?

Reply:
I think in the admin, the model's `save` method is called directly, which explains the above error, but overriding the save method on the model does not work either (and is also not recommended in the Django docs). Maybe I'm overlooking a simple way of doing this, but if not, maybe raising an error like I did before in the admin is the way to go after all.

Reply:
Maybe lets save ourself some time - do we really need to do this in the admin? We could treat this as a no go. When would we need to create interfaces without going through an algorithm? Any algorithm editor can do it themselves if they need to (add algorithm perms), and it would be done automatically for challenge participants. The only other place I can think of doing this would be for phases but again a standard form will do. Best to pick our battles.

Reply:
Ok. The only use case I see is someone without the add_algorithm permission wanting to host an alogirhtm outside of a challenge. They pay us and so far, I think, we did the basic set-up for them. That would no longer work if we can't create interfaces in the admin. But it would probably be better anyway to grant this person add_algorithm permission through a special group membership?

Reply:
Forget what I say above: we could create the algorithm and its interfaces for them through the UI, of course. Duh.

Reply:
I'll remove the add ability from the admin views then